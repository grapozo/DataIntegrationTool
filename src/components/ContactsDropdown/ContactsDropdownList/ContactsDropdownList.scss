// NOTE: For convenience, we import the dropdown configuration variables here.
// If this component is moved in the future, we'll need to redeclare these values.
@use '../configuration/DropdownConfigurations' as *;

.contactsDropdownList {
  text-align: left;
  overflow-y: scroll;
  // The max height uses the figma design as a base.
  max-height: 5rem;
  opacity: 0;
  visibility: hidden;
  // Animation for hiding/showing the component
  transition: visibility $animation-speed linear, opacity $animation-speed linear;

  // Needed to keep the figma-design place of the scrollbar.
  margin-right: 0.3125rem;

  &.isVisible {
    // Shows the component
    opacity: 1;
    visibility: visible;
  }

  // Chrome scrollbar
  &::-webkit-scrollbar {
    width: 0.1875rem;
    margin-right: 1rem;
  }

  &::-webkit-scrollbar-thumb {
    background: #cbd6e2;
    border-radius: 50px;
  }

  // Firefox scrollbar
  scrollbar-color: #cbd6e2 #f5f8fa;
  scrollbar-width: thin;
}
