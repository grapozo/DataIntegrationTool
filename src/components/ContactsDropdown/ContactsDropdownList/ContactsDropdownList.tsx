import { useEffect, useState } from 'react';
import { Contact } from 'shared/interfaces/contact.interface';
import styles from './ContactsDropdownList.scss';
import { ContactDropdownListItem } from './ContactsDropdownListItem';

type ContactsDropdownListProps = {
  /** True if the DropdownList should be visible. */
  isVisible: boolean;
  /** The contacts list */
  contacts: Contact[];
  /** Callback that runs when the selected contact list is changed */
  selectedContactListChangedCallback: (selectedContacts: Contact[]) => void;
};

export const ContactsDropdownList = ({
  isVisible,
  contacts,
  selectedContactListChangedCallback
}: ContactsDropdownListProps): JSX.Element => {
  const [selectedContacts, setSelectedContacts] = useState<Contact[]>([]);

  /**
   * Add/Removes a selected contact from the selectedContacts array.
   * @param contact The selected contact
   * @param contactIsSelected Will be true if contact is selected
   */
  const handleContactSelect = (contact: Contact, contactIsSelected: boolean): void => {
    if (contactIsSelected) {
      // Add contact to the selected list.
      setSelectedContacts((contacts) => [...contacts, contact]);
    } else {
      // ... Removes the contact from the list.
      setSelectedContacts((contacts) =>
        contacts.filter((contactItr) => contactItr.id !== contact.id)
      );
    }
  };

  // Notifies parent of a Selected Contact change.
  useEffect(() => {
    selectedContactListChangedCallback(selectedContacts);
  }, [selectedContacts]);

  return (
    <div
      className={`
      ${styles.contactsDropdownList} 
      ${isVisible ? styles.isVisible : ''}`}
    >
      {contacts.map(function (contact) {
        return (
          <div key={contact.id}>
            <ContactDropdownListItem
              contact={contact}
              onContactSelectCallback={handleContactSelect}
            />
          </div>
        );
      })}
    </div>
  );
};
