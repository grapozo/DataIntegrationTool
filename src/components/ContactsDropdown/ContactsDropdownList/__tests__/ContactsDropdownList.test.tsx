import { create } from 'react-test-renderer';
import { Contact } from 'shared/interfaces/contact.interface';
import { ContactsDropdownList } from '../ContactsDropdownList';
import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

const mockContactList: Contact[] = [
  {
    id: '9ba3f8f7d420c5fb33d7fbf2704e24fe',
    name: 'Mock contact'
  },
  {
    id: '5a87a7431aafb3def979500d332eea05',
    name: 'Mock contact 2'
  }
];

const mockSelectedContactListChangedCallback = jest.fn();

const MockDropdownList = (): JSX.Element => {
  return (
    <ContactsDropdownList
      contacts={mockContactList}
      isVisible={true}
      selectedContactListChangedCallback={mockSelectedContactListChangedCallback}
    />
  );
};
/**
 * In this component, testing some edge cases may be difficult, because
 * some functionality is relying on classNames.
 */
describe('ContactDropdownListItem', () => {
  it('should correctly render', () => {
    const tree = create(<MockDropdownList />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should call selectedContactListChangedCallback when a contact is selected/unselected', () => {
    render(<MockDropdownList />);
    // Selects the first contact on the list
    fireEvent.click(screen.getByLabelText(mockContactList[0].name));
    expect(mockSelectedContactListChangedCallback).toHaveBeenCalledWith([mockContactList[0]]);
    // Unselect the first contact on the list
    fireEvent.click(screen.getByLabelText(mockContactList[0].name));
    expect(mockSelectedContactListChangedCallback).toHaveBeenCalledWith([]);
  });
  it('should correctly render the contacts list', async () => {
    render(<MockDropdownList />);
    const checkboxes = await screen.findAllByRole('checkbox');
    expect(checkboxes.length).toBe(mockContactList.length);
  });
});
