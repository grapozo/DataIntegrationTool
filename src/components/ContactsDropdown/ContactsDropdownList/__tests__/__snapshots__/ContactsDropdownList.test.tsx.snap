// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactDropdownListItem should correctly render 1`] = `
<div
  className="
      undefined 
      undefined"
>
  <div>
    <div
      className="undefined"
    >
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        Mock contact
      </label>
    </div>
  </div>
  <div>
    <div
      className="undefined"
    >
      <label>
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        Mock contact 2
      </label>
    </div>
  </div>
</div>
`;
